pkg/controllermanager/controller/project/project_control_reconcile.go

```diff
@@ -81,15 +80,15 @@ func (r *projectReconciler) reconcile(ctx context.Context, project *gardencorev1
 
 	// Update the name of the created namespace in the projects '.spec.namespace' field.
 	if ns := project.Spec.Namespace; ns == nil {
-		if err := kutil.TryPatch(ctx, retry.DefaultBackoff, gardenClient, project, func() error {
-			project.Spec.Namespace = &namespace.Name
-			return nil
-		}); err != nil {
+		project.Spec.Namespace = &namespace.Name
+		if err := gardenClient.Update(ctx, project); err != nil {
```

Comment:
Why is `Update` better than `TryPatch` here (and below)? It can still lead to conflicts, and now there is no attempt to retry?

Reply:
Excellent question! This is an important question, I definitely want to answer in the developer guideline on the clients (started to work on it this morning). Nevertheless, here is my reasoning behind this particular case:

If we encounter a conflict when updating the `spec.namespace` field (or the ownerRefs on the namespace in the case below), it means that we potentially operated on stale data and should retry the entire reconcile operation from the beginning.
For example, if the project has `spec.namespace` set in the meantime and we retry on conflict, we blindly overwrite the field and thereby orphan a namespace. Instead, previous `if`s should have evaluated differently and we should have taken different actions (i.e. adopt the namespace instead of creating one). 
This means, we shouldn't simply ignore the conflict here and explicitly make use of it (by using update or patch with optimistic locking) and not retrying, because it's actually a good thing, that prevents us from operating on stale data and issuing wrongful concurrent resource updates.

So the general statement would be: conflicts are not generally something we want to avoid. In some cases it is important to leverage the built-in concurrency prevention mechanisms and handle conflicts properly by relying on the controller's exponential backoff.
In most cases however, we want to ignore concurrent update to resources, as we are the exclusive owner of the set of fields we are updating (e.g. spec of extension resources or controlplane deployments). But the correct way to accomplish that is by making precise updates to only these fields (i.e. by using patches) and running without optimistic locking.
But in both cases `retryOnConflict` is not the right option to solve the situation (there are some correct use cases for it, though they are very rare).

I hope, this was helpful and understandable :)

(Side-note: the controller is still not perfect and can still lead to inconsistent states, as noted in the comment below. It should be further improved. But the usage of the retryOnConflict-semantic is definitely wrong here, so I replaced it)

Reply:
Thanks a lot for the great explanation! Yes, we definitely need a guideline when to use what update / patch mechanism, I would be happy to take a look when you have it.